website/blog/2024-08-06-langgraph-pyfunc/index.md

```diff
@@ -0,0 +1,223 @@
+---
+title: LangGraph with Custom PyFunc 
+tags: [genai, mlops]
+slug: mlflow-tracing
+authors: [mlflow-maintainers]
+thumbnail: img/blog/release-candidates.png
+---
+
+In this blog, we'll guide you through creating a LangGraph chatbot within an MLflow custom PyFunc.
+
+### What is a Custom PyFunc?
+
+While MLflow strives to cover the many popular machine learning libraries, there are libraries and pieces of functionality within supported libraries that are not natively supported. For these cases, when users want MLflow functionality they can create a custom PyFunc model. 
+Custom PyFunc models allow you to integrate any Python code, providing flexibility in defining complex and niche machine learning models. These models can be easily logged, managed, and deployed using the typical MLflow APIs, enhancing flexibility and portability in machine learning workflows.
+
+### What is LangGraph?
+
+LangGraph is a library for building stateful, multi-actor applications with LLMs, used to create agent and multi-agent workflows. Compared to other LLM frameworks, it offers these core benefits: cycles, controllability, and persistence. LangGraph allows you to define flows that involve cycles, essential for most agentic architectures, differentiating it from DAG-based solutions. As a very low-level framework, it provides fine-grained control over both the flow and state of your application, crucial for creating reliable agents. Additionally, LangGraph includes built-in persistence, enabling advanced human-in-the-loop and memory features.
+LangGraph is inspired by Pregel and Apache Beam. The public interface draws inspiration from NetworkX. LangGraph is built by LangChain Inc, the creators of LangChain, but can be used without LangChain.
+
+# The Code
+
+### Setup
+First, we must install the required dependencies. We will use OpenAI for our LLM, but LangChain paired with LangGraph makes it easy to substitute your desired LLM.
+
+```python
+%%capture --no-stderr
```

Comment:
Is this super chatty? 
Typically I put statements like this that send warnings / messages to devnull void of dispair only when doing a screen recording of running something. This might look fishy to leave this in.